infobind/cache.go

```diff
@@ -0,0 +1,204 @@
+type BindCacheUpdater struct {
+	ctx sessionctx.Context
+
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	globalHandler  *Handler
+}
+
+type bindMeta struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	// Status will only be 0 or 1:
+	//   0: bindMeta has been marked as deleted status.
+	//   1: bindMeta is in using status.
+	Status     int64
+	CreateTime types.Time
+	UpdateTime types.Time
+	Charset    string
+	Collation  string
+}
+
+// NewBindCacheUpdater create a new BindCacheUpdater.
+func NewBindCacheUpdater(ctx sessionctx.Context, handler *Handler, parser *parser.Parser) *BindCacheUpdater {
+	return &BindCacheUpdater{
+		globalHandler: handler,
+		parser:        parser,
+		ctx:           ctx,
+	}
+}
+
+// NewHandler create a Handler with a bindCache.
+func NewHandler() *Handler {
+	handler := &Handler{}
+	return handler
+}
+
+// Get get bindCache from a Handler.
+func (h *Handler) Get() bindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(map[string][]*bindRecord)
+	}
+
+	return make(map[string][]*bindRecord)
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *BindCacheUpdater) loadDiff(sql string, bc bindCache) error {
+	recordSets, err := h.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chkBatch.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := decodeBindTableRow(row)
+			err = bc.appendNode(record, h.parser)
+			if err != nil {
+				return err
+			}
+
+			if record.UpdateTime.Compare(h.lastUpdateTime) == 1 {
+				h.lastUpdateTime = record.UpdateTime
+			}
+		}
+	}
+}
+
+// Update update the BindCacheUpdater's bindCache if tidb first startup,the fullLoad is true,otherwise fullLoad is false.
+func (h *BindCacheUpdater) Update(fullLoad bool) (err error) {
+	var sql string
+	bc := h.globalHandler.Get()
+	newBc := make(map[string][]*bindRecord, len(bc))
+	for hash, bindDataArr := range bc {
+		newBc[hash] = append(newBc[hash], bindDataArr...)
+	}
+
+	if fullLoad {
+		sql = "select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info"
+	} else {
+		sql = fmt.Sprintf("select original_sql, bind_sql, default_db, status, create_time, update_time, charset, collation from mysql.bind_info where update_time > \"%s\"", h.lastUpdateTime.String())
+	}
+	err = h.loadDiff(sql, newBc)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.globalHandler.bind.Store(newBc)
+
+	return nil
+}
+
+func decodeBindTableRow(row chunk.Row) bindMeta {
```

Comment:
1. s/decodeBindTableRow/newBindMeta/
2. How about returning a `bindMeta` pointer, i.e. change the return value to `*bindMeta`? Thus we can save one copy.